lock-keeper-client-cli/src/cli_command/mod.rs

```diff
@@ -0,0 +1,127 @@
+    async fn execute(self: Box<Self>, state: &mut State) -> Result<(), anyhow::Error>;
+
+    /// Given the proper string arguments to a command, return an instance of
+    /// this command.
+    fn parse_command_args(args: &[&str]) -> Option<Self>
+    where
+        Self: Sized;
+
+    /// Expected format for this command such as "register [account_name]
+    /// [password]". Used for generating help string dynamically.
+    fn format() -> &'static str
+    where
+        Self: Sized;
+
+    fn aliases() -> Vec<&'static str>
+    where
+        Self: Sized;
+
+    fn description() -> &'static str
+    where
+        Self: Sized;
+
+    /// Convert a string into a Box<dyn Command> suitable for dynamically
+    /// dispatching the execute function.
+    fn to_dyn(self) -> DynCommand
+    where
+        Self: Sized + 'static,
+    {
+        Box::new(self)
+    }
+
+    fn from_str(s: &str) -> Result<Self, anyhow::Error>
+    where
+        Self: Sized,
+    {
+        let mut split: _ = s.trim().split(' ');
+
+        let command = split
+            .next()
+            .ok_or_else(|| anyhow!("Missing \"command\" part of argument."))?;
+        debug!("Looking for match for \"{}\"", command);
+
+        // Check if this command matches any of our aliases. Otherwise return.
+        // We do this check here to avoid doing it inside every
+        // `CliCommand::parse_command_args` function.
+        if !Self::aliases().contains(&command) {
+            info!("Valid aliases for this command are: {:?}", Self::aliases());
+            bail!("Command does not match any know name or alias.");
+        }
+        info!("Found valid command: {}", command);
+
+        match Self::parse_command_args(&split.collect::<Vec<_>>()) {
+            Some(t) => Ok(t),
+            None => bail!("Expected Format: {}", Self::format()),
+        }
+    }
+}
+
+// We list all our command types here. We can generate the correct functions by
+// implementing GetFunction.
+// If you implement a new CliCommand, please add it to this list.
+pub fn get_functions<F: GetCliCommandFunctions>() -> Vec<F::Function> {
+    vec![
+        F::get_functions::<Authenticate>(),
+        F::get_functions::<Quit>(),
+        F::get_functions::<Export>(),
+        F::get_functions::<Generate>(),
+        F::get_functions::<GetAuditEvents>(),
+        F::get_functions::<Help>(),
+        F::get_functions::<Import>(),
+        F::get_functions::<List>(),
+        F::get_functions::<Logout>(),
+        F::get_functions::<Register>(),
+        F::get_functions::<Register>(),
+        F::get_functions::<RemoteGenerate>(),
+        F::get_functions::<RemoteSign>(),
+        F::get_functions::<Retrieve>(),
+    ]
+}
+
+/// The GetCliCommandFunctions trait allows us to generate a vector of the same
+/// function for all our commands. See [get_functions] for use.
+pub trait GetCliCommandFunctions {
+    type Function;
+
+    fn get_functions<T: CliCommand + 'static>() -> Self::Function;
```

Comment:
```suggestion
    fn get_function<T: CliCommand + 'static>() -> Self::Function;
```

I think this should be singular since it only returns one function.